cli: default --include-range-info to true, but create a single file

tbg opened the pull request:
This was introduced and defaulted to `false` in
https://github.com/cockroachdb/cockroach/pull/102289.

Anecdotally I look at these in many L2 escalations[^1], and one cannot always
even change the default flag (e.g. CC), and seldom will a later debug.zip be
good enough for an RCA.

The original reason to introduce this flag seems to have been the delay
incurred when extracting these many files. There are suspicions that this
somehow exacerbates broken archives that we get from time to time, esp.
on large clusters.

This PR defaults the flag to true but also switches debug.zip so that it now
puts the ranges into a single file. This file is going to be large, but we can
handle that.

[^1]: most recently today: https://github.com/cockroachlabs/support/issues/2343#issuecomment-1569677115

Epic: none
Release note: None

tbg committed:
cli: default --include-range-info to true

This was introduced and defaulted to `false` in
https://github.com/cockroachdb/cockroach/pull/102289.

Anecdotally I look at these in many L2 escalations[^1], and one cannot always
even change the default flag (e.g. CC), and seldom will a later debug.zip be
good enough for an RCA.

The original reason to introduce this flag seems to have been the delay
incurred when extracting these many files, but they can be skipped via `unzip
-x "debug/nodes/*/ranges/*"`.

Besides, the flag is still there, so we can ask operators of large deployments
to opt in selectively.

[^1]: most recently today: https://github.com/cockroachlabs/support/issues/2343#issuecomment-1569677115

Epic: none
Release note: None

erikgrinaker reviewed (approved):
+1, I often look at these too, and if they're not included it's often too late to ask the operator for a new debug.zip.

dt commented:
> concat them into a single file per node instead

I'm also happy to volunteer to do this today or tomorrow if this is the direction we want to go.

tbg commented:
> In practice, most of the zips we ge from the biggest clusters end up incompatible with unzip, requiring us to use other utilities (e.g. finder) that don't have the option to exclude paths.

Do you have more information on this? This is news to me. I saw [this thread](https://cockroachlabs.slack.com/archives/C01CNRP6TSN/p1684265465790619) but this didn't seem particular to how the files were arranged: it sounded more like large zip files sometimes get uploaded only partially.

> concat them into a single file per node instead

Fine with that, though I'm not sure it makes the files more "compatible" with unzip, sure can't hurt and it's easier to handle overall.

> I'm also happy to volunteer to do this today or tomorrow if this is the direction we want to go.

Let me see if I can do this in 10 minutes and I'll do it here.

tbg committed:
cli: write range jsons as single file

Epic: none
Release note (ops change): The debug.zip archive now contains the files formerly
located at `nodes/*/ranges/*.json` in one file per node, `nodes/*/ranges.json`.

dt commented on pkg/cli/zip_per_node.go:
```diff
@@ -529,17 +528,10 @@ func (zc *debugZipContext) collectPerNodeData(
-				}
+			s := nodePrinter.start("writing ranges")
+			name := fmt.Sprintf("%s/ranges.json", prefix)
+			if err := zc.z.createJSON(s, name, ranges.Ranges); err != nil {
```
Not suggesting, just asking: Do we think we should marshal each individually, writing the results to a writer opened above the loop as we go, rather than marshaling all 100k at once? how big is one range json?

tbg committed:
cli: use AssertHeld in debug.zip

Epic: none
Release note: None

tbg committed:
cli: don't buffer zip JSON in memory

Epic: none
Release note: None

tbg committed:
cli: add missing fail() call in zip

Epic: none
Release note: None

tbg commented:
TFTR!

bors r=dt
